docs/blueprints/referral-manager-v1.md

```diff
@@ -0,0 +1,232 @@
+2. Referral Manager checks the status of the token:
+    - if the token exists in `tokens` table:
+        - if it is not redeemed, Referral Manager sends back a success response to the redeemed satellite and mark the token as redeemed in the Referral Manager's database
+        - if it token is already redeemed, Referral Manager sends back a `invalid token` response to the redeemed satellite.
+        
+    - if the token doesn't exist in `tokens` table, Referral Manager sends back a `invalid token` response to the redeemed satellite.
+3. Redeemed satellite receives the response, which:
+    - if it is a success, the satellite will proceed with the account creation and
+    then send a request to the Referral Manager, which: 
+        - updates the `tokens` table to store the newly created user ID in `redeemedID` column for that particular referral token.
+        - updates the `users` table to store the newly created user ID and the satellite URL the user belongs to.
+    - if it is an invalid token, the satellite will display a proper message in the UI.
+    
+## Rationale
+
+We will create a `inspector` port that's only listening on `localhost` so CLI could be only used by Referral Manager operators.
+
+We could also set the `inspector` port to only accept requests coming from storj vpn.
+
+## Implementation
+
+- Create a private repository for Referral Manager.
+- Create `tokens` and `users` table in Referral Manager database.
+- Create an `inspector` endpoint on Referral Manager.
+    - Implementing a method `GenerateTokens` for generating referral tokens and save them into `tokens` table on Referral Manager.
+    - Implementing a mathod `GetEligibleUsers` for querying users who don't have any referral tokens associated with.
+- Implementing an endpoint `GetTokens` for requesting unredeemed tokens from Referral Manager. 
+- Implementing an endpoint `Redeem` on Referral Manager for verifying invitation tokens.
+- Implementing an endpoint `ClearUsers` on Referral Manager for setting in-memory `eligibleUsers` to be empty from Referral Manager CLI.
+- Replace existing registration token logic.
+
+### Pseudocode
+
+The Token struct represents both a Go struct (on satellite and Referral Manager) as well as the schema for the `tokens` table on the Referral Manager
+
+```
+type Token struct {
+    Secret [32]byte
+    OwnerID uuid.UUID
+    RedeemedID uuid.UUID
+    OwnerSatelliteURL string
+    RedeemedSatelliteURL string
+    Status string
+}
+```
+Statuses: `unsent` (owner's satellite doesn't know about it yet), `unredeemed` (owner's satellite knows about it but it has not been used), `redeemed` (someone has used this referral link to register alreadys)
+
+The User struct represent both a Go struct as well as the schema for the `users` table on Referral Manager.
+
+```
+type User struct {
+    ID uuid.UUID
+    satelliteURL string
+}
+```
+
+
+**Endpoints for satellites:**
```

Comment:
We should also add `RegisterUser` here

Reply:
Actually, should that be a part of `Redeem`? Maybe we don't need a separate `RegisterUser` endpoint.

Reply:
Yea, i imagined it to be a part of the `Redeem`

Reply:
Added the pseudocode